pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -242,90 +317,276 @@ func (sf *IteratorDatastore) Read(
+	delete(sf.internalStorage.iters, cacheKey)
+	sf.internalStorage.mu.Unlock()
+	item.iter.cleanup()
 }
 
 // sharedIterator will be shared with multiple consumers.
-// For now, this is a skeleton and no one is actually sharing the iterator.
 type sharedIterator struct {
-	manager *IteratorDatastore
-	key     string
-	inner   storage.TupleIterator
-	stopped atomic.Bool
+	manager      *IteratorDatastore // non-changing
+	key          string             // non-changing
+	head         int
+	maxAliveTime time.Duration
+	queryErr     error
+
+	mu                 *sync.Mutex           // We expect the contention should be minimal. TODO: minimize mu holding time.
+	items              *[]*openfgav1.Tuple   // shared - protected by mu
+	inner              storage.TupleIterator // shared - protected by mu
+	sharedErr          *error                // shared - protected by mu
+	watchdogTimer      *time.Timer           // shared - protected by mu
+	watchdogTimeoutErr *error                // shared - protected by mu
+	stopped            bool                  // not shared across clone - but protected by mu
+}
+
+func newSharedIterator(manager *IteratorDatastore, key string, maxAliveTime time.Duration) *sharedIterator {
+	newIter := &sharedIterator{
+		manager:      manager,
+		key:          key,
+		head:         0,
+		maxAliveTime: maxAliveTime,
+
+		mu:                 &sync.Mutex{},
+		items:              new([]*openfgav1.Tuple),
+		inner:              nil,
+		sharedErr:          new(error),
+		watchdogTimeoutErr: new(error),
+	}
+	newIter.watchdogTimer = time.AfterFunc(maxAliveTime, newIter.watchdogTimeout)
+	return newIter
+}
+
+// It is assumed that mu is held by the parent while cloning.
+func (s *sharedIterator) clone() (*sharedIterator, error) {
+	if s.queryErr != nil {
+		return nil, s.queryErr
+	}
+	newIter := &sharedIterator{
+		manager:      s.manager,
+		key:          s.key,
+		head:         0,
+		maxAliveTime: s.maxAliveTime,
+
+		mu:                 s.mu,
+		items:              s.items,
+		inner:              s.inner,
+		sharedErr:          s.sharedErr,
+		watchdogTimer:      s.watchdogTimer,
+		watchdogTimeoutErr: s.watchdogTimeoutErr,
+	}
+	newIter.watchdogTimer.Reset(s.maxAliveTime)
+	return newIter, nil
 }
 
-func newSharedIterator(manager *IteratorDatastore, key string, inner storage.TupleIterator) *sharedIterator {
-	return &sharedIterator{
-		manager: manager,
-		key:     key,
-		inner:   inner,
+func (s *sharedIterator) cleanup() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.inner != nil {
+		s.inner.Stop()
 	}
+	s.watchdogTimer.Stop()
+}
+
+// when watchdogTimeout is invoked, it will revoke the current iterator from usage.
+// This will mark the iterator as error and stop on its behalf
+// All further usage of Head/Next() will return error.
+func (s *sharedIterator) watchdogTimeout() {
+	s.mu.Lock()
+	*s.watchdogTimeoutErr = errSharedIteratorWatchdog
+	s.mu.Unlock()
+	sharedIteratorWatchDog.Inc()
+	s.manager.watchdogTimeout(s.key)
```

Comment:
Consider the following case:
```
watchdog timeout scheduled
deref last item.  Which cause the manger to remove the item
Watchdog run
```

In this case, it is possible to have two scenarios:
1. No entry is created.  In this case, it is ok because when we deref the last item, we would have all the associated iter / etc.
2. New entry is created between deref last item &&  - this is bad as we have prematurely remove the entry and associated info.  

Consider adding some form of unique identifier to avoid case #2